chain/vm/fvm.go

```diff
@@ -0,0 +1,310 @@
+			log.Info("invalid consensus fault: cannot decode extra: %w", decodeErr)
+			return ret, totalGas
+		}
+
+		if types.CidArrsEqual(blockA.Parents, blockC.Parents) && blockA.Height == blockC.Height &&
+			types.CidArrsContains(blockB.Parents, blockC.Cid()) && !types.CidArrsContains(blockB.Parents, blockA.Cid()) {
+			faultType = ffi_cgo.ConsensusFaultParentGrinding
+		}
+	}
+
+	// (3) return if no consensus fault by now
+	if faultType == ffi_cgo.ConsensusFaultNone {
+		log.Info("invalid consensus fault: no fault detected")
+		return ret, totalGas
+	}
+
+	// else
+	// (4) expensive final checks
+
+	// check blocks are properly signed by their respective miner
+	// note we do not need to check extra's: it is a parent to block b
+	// which itself is signed, so it was willingly included by the miner
+	gasA, sigErr := x.VerifyBlockSig(ctx, &blockA)
+	totalGas += gasA
+	if sigErr != nil {
+		log.Info("invalid consensus fault: cannot verify first block sig: %w", sigErr)
+		return ret, totalGas
+	}
+
+	gas2, sigErr := x.VerifyBlockSig(ctx, &blockB)
+	totalGas += gas2
+	if sigErr != nil {
+		log.Info("invalid consensus fault: cannot verify second block sig: %w", sigErr)
+		return ret, totalGas
+	}
+
+	ret.Type = faultType
+	ret.Target = blockA.Miner
+
+	return ret, totalGas
+}
+
+func (x *FvmExtern) VerifyBlockSig(ctx context.Context, blk *types.BlockHeader) (int64, error) {
+	waddr, gasUsed, err := x.workerKeyAtLookback(ctx, blk.Miner, blk.Height)
+	if err != nil {
+		return gasUsed, err
+	}
+
+	return gasUsed, sigs.CheckBlockSignature(ctx, blk, waddr)
+}
+
+func (x *FvmExtern) workerKeyAtLookback(ctx context.Context, minerId address.Address, height abi.ChainEpoch) (address.Address, int64, error) {
+	gasUsed := int64(0)
+	gasAdder := func(gc GasCharge) {
+		// technically not overflow safe, but that's fine
+		gasUsed += gc.Total()
+	}
+
+	cstWithoutGas := cbor.NewCborStore(x.Blockstore)
+	cbb := &gasChargingBlocks{gasAdder, PricelistByEpoch(x.epoch), x.Blockstore}
+	cstWithGas := cbor.NewCborStore(cbb)
+
+	lbState, err := x.lbState(ctx, height)
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+	// get appropriate miner actor
+	act, err := lbState.GetActor(minerId)
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+
+	// use that to get the miner state
+	mas, err := miner.Load(adt.WrapStore(ctx, cstWithGas), act)
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+
+	info, err := mas.Info()
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+
+	stateTree, err := state.LoadStateTree(cstWithoutGas, x.base)
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+
+	raddr, err := ResolveToKeyAddr(stateTree, cstWithGas, info.Worker)
+	if err != nil {
+		return address.Undef, gasUsed, err
+	}
+
+	return raddr, gasUsed, nil
+}
+
+type FVM struct {
+	fvm *ffi.FVM
+}
+
+func NewFVM(ctx context.Context, opts *VMOpts) (*FVM, error) {
+	circToReport := opts.FilVested
```

Comment:
Are there no solutions other than snaking through this `FilVested` field? Seems a little bit ad-hoc when combined with the pre-existing `CircSupplyCalc` field. And the fact that it won't be even used nv15 and forward, makes me even uneasier.

An alternative solution would be to expand the signature of `CircSupplyCalc` to take in the network version. Since the value of that is a `StateManager` member, it has all it needs to return `FilVested` if nv < 15.

Then this logic can be entirely independent of the network version, and we'd just pass the value down and let the FVM do the right thing.

Reply:
@raulk My priority here is not changing the existing logic for the Lotus VM. I would rather have a new field that the LotusVM entirely ignores -- that way if there's anything wrong with this PR, only FVM M0 testing is affected.

If it'll make you feel better, we can drop this entirely from the final released version of Lotus for the v16 upgrade and just always use the calculated CircSupply.